azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_artifactory.go

```diff
@@ -0,0 +1,172 @@
+		MaxItems:     1,
+		Elem:         at,
+		ExactlyOneOf: []string{"authentication_password", "authentication_token"},
+	}
+
+	r.Schema["authentication_password"] = &schema.Schema{
+		Type:     schema.TypeList,
+		Optional: true,
+		MinItems: 1,
+		MaxItems: 1,
+		Elem:     aup,
+	}
+
+	return r
+}
+
+// Convert internal Terraform data structure to an AzDO data structure
+func expandServiceEndpointArtifactory(d *schema.ResourceData) (*serviceendpoint.ServiceEndpoint, *string, error) {
+	serviceEndpoint, projectID := doBaseExpansion(d)
+	serviceEndpoint.Type = converter.String("artifactoryService")
+	serviceEndpoint.Url = converter.String(d.Get("url").(string))
+	authScheme := "Token"
+
+	authParams := make(map[string]string)
+
+	if x, ok := d.GetOk("authentication_token"); ok {
+		authScheme = "Token"
+		msi := x.([]interface{})[0].(map[string]interface{})
+		authParams["apitoken"] = expandSecret(msi, "access_token")
+	} else if x, ok := d.GetOk("authentication_password"); ok {
+		authScheme = "UsernamePassword"
+		msi := x.([]interface{})[0].(map[string]interface{})
+		authParams["username"] = expandSecret(msi, "username")
+		authParams["password"] = expandSecret(msi, "password")
+	}
+	serviceEndpoint.Authorization = &serviceendpoint.EndpointAuthorization{
+		Parameters: &authParams,
+		Scheme:     &authScheme,
+	}
+
+	return serviceEndpoint, projectID, nil
+}
+
+func expandSecret(credentials map[string]interface{}, key string) string {
+	// Note: if this is an update for a field other than `key`, the `key` will be
+	// set to `""`. Without catching this case and setting the value to `"null"`, the `key` will
+	// actually be set to `""` by the Azure DevOps service.
+	//
+	// This step is critical in order to ensure that the service connection can update without loosing its password!
+	//
+	// This behavior is unfortunately not documented in the API documentation.
+	val, ok := credentials[key]
+	if !ok || val.(string) == "" {
+		return "null"
+	}
+
+	return val.(string)
+}
+
+// Convert AzDO data structure to internal Terraform data structure
+func flattenServiceEndpointArtifactory(d *schema.ResourceData, serviceEndpoint *serviceendpoint.ServiceEndpoint, projectID *string) {
+	doBaseFlattening(d, serviceEndpoint, projectID)
+	if strings.EqualFold(*serviceEndpoint.Authorization.Scheme, "Token") {
+		auth := make(map[string]interface{})
+		if x, ok := d.GetOk("authentication_token"); ok {
+			authList := x.([]interface{})[0].(map[string]interface{})
+			if len(authList) > 0 {
+				newHash, hashKey := tfhelper.HelpFlattenSecretNested(d, "authentication_token", authList, "access_token")
+				auth[hashKey] = newHash
+			}
+		}
+		if serviceEndpoint.Authorization != nil && serviceEndpoint.Authorization.Parameters != nil {
+			auth["access_token"] = (*serviceEndpoint.Authorization.Parameters)["apitoken"]
+		}
+		d.Set("authentication_token", []interface{}{auth})
+	}
+	if strings.EqualFold(*serviceEndpoint.Authorization.Scheme, "UsernamePassword") {
```

Comment:
I would suggest change `if` to `else if`

Reply:
Done